### PR TITLE
[wicket] Start to pull real inventory from wicketd

### PR DESCRIPTION
**Pull request by andrewjstone**
This commit hooks wicket up to wicketd via wicketd-client. Wicket no longer publishes fake inventory, and no longer plans to interact with MGS. Instead it only talks to wicketd which acts as a caching proxy for MGS inventory.

The data retrieved from wicketd is converted into types already used by the display screen infrastructure in wicket. This is what's happening in `inventory.rs`. When the inventory is updated the screen is redrawn.

This is just a start, and quite a few things are missing:
 * Real power state - this is currently faked to A2 for everything
 * ComponentInfo (actual SP inventory)
 * Version info - This still needs to be implemented on the SPs and plumbed all the way up
 * A display for the last time wicketd was talked to and if MGS is actually alive. We want to show how stale our data is. Right now if wicketd goes down we'll log it but not notice in the UI. If MGS goes down we don't have any idea.
 * ETAG support and diff support so we don't redraw the same inventory over and over.

This was tested via running MGS, wicketd and sp simulators on my local helios box. If you run wicket standalone it still works fine, but the component screen won't show any inventory.

### CODE REVIEW

**Comment by andrewjstone** [comment]
Here's a preview gif of publishing data:Screen Recording 2022-11-11 at 6 29 33 PM

File: wicket/src/inventory.rs [hunk 1]
```diff
@@ -20,94 +21,70 @@ impl Inventory {
...
-    pub sp_version: String,
-    pub rot_version: String,
+pub struct Sp {
+    ignition: SpIgnition,
+    state: SpState,
 }
 
-/// TODO: Use real inventory received from MGS
+// XXX: Eventually a Sled will have a host component, and SPs will have
+// `SpComponentInfo` (device inventory) [derive(Debug)]
```
Comment (jgallagher) [inline wicket/src/inventory.rs:82]:
```suggestion
// `SpComponentInfo` (device inventory)
```

File: wicket/src/inventory.rs [hunk 2]
```diff
@@ -20,94 +21,70 @@ impl Inventory {
...
+                    (ComponentId::Switch(i as u8), Component::Switch(sp))
+                }
+                SpType::Power => {
+                    if i > 1 {
+                        return Err(anyhow!("Invalid power shelf slot: {}", i));
+                    }
+                    (ComponentId::Psc(i as u8), Component::Psc(sp))
+                }
+            };
+            new_inventory.inventory.insert(id, component);
```
Comment (jgallagher) [inline wicket/src/inventory.rs:60]: It's a little weird that `Inventory` has a field also named `inventory`, I think? This question might not make sense in a larger context, but why do we want to record the component info and power state in separate fields?

File: wicket/src/wicketd.rs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+use crate::Event;
+use slog::{o, warn, Logger};
+use std::net::SocketAddrV6;
+use std::sync::mpsc::Sender;
+use tokio::sync::mpsc;
+use tokio::time::{interval, Duration, MissedTickBehavior};
+use wicketd_client::types::RackV1Inventory;
+
+const WICKETD_POLL_INTERVAL: Duration = Duration::from_secs(5);
+const WICKETD_TIMEOUT_MS: u32 = 1000; // 3 sec
```
Comment (jgallagher) [inline wicket/src/wicketd.rs:16]: Either the comment or the value is wrong here :)

File: wicket/src/wicketd.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+// Assume that these requests are periodic on the order of seconds or the
+// result of human interaction. In either case, this buffer should be plenty
+// large.
+const CHANNEL_CAPACITY: usize = 1000;
```
Comment (jgallagher) [inline wicket/src/wicketd.rs:21]: This is probably a question for a later PR (I don't think this is actually used yet, right?), but:

a) What happens if the channel fills completely?
b) What's the effect if the channel isn't completely full but has several hundred items in it?

If the answer to (a) is "the UI becomes unresponsive" and the answer to (b) is "the UI remains responsive but is very out of date", I'm not sure that's a whole lot better. Maybe a smaller channel with a way to politely block the UI or give the user a way to cancel outstanding requests? That sounds pretty complicated already, so maybe disregard. Just wanted to dump some thoughts while I was reading.

**Review by jgallagher — APPROVED** [review]

File: wicket/src/inventory.rs (hunk 2 above)
Comment (andrewjstone) [reply wicket/src/inventory.rs:60]: Ah, so this is just a historical artifact. I wanted to plug in the real inventory without changing the UI at this time. I think it maybe worth a rework at some point, possibly as I pull in real power.

File: wicket/src/wicketd.rs (hunk 4 above)
Comment (andrewjstone) [reply wicket/src/wicketd.rs:21]: These are great questions. As you noted the channel isn't in use yet, but will be soon. I opened an issue to track: https://github.com/oxidecomputer/omicron/issues/1942

Comment (andrewjstone) [reply wicket/src/wicketd.rs:21]: We've been using wicket for about a year now and have not seen an issue here. Closing this.

**Pull request merged**